### PR TITLE
fix time_bucket_tz(day): don't truncate input to a day boundary

### LINKED ISSUE #9260: time_bucket selects wrong bin for day intervals with unaligned origins.
### What happens?

when binning using time_bucket and a day interval the bin is incorrect if the origin is not day aligned. verified that postgresql, when using date_bin, produces the expected result.

### To Reproduce

both columns should be equal -- the tstz is wrong.
```sql
set TimeZone='UTC';
select
  time_bucket(interval '1' day, timestamptz '2023-10-07 16:08:09+00', origin) tstz,
  time_bucket(interval '1' day, timestamp   '2023-10-07 16:08:09', origin::timestamp) ts
from
  generate_series(timestamptz '2001-01-03 00:00:00+00',
                  timestamptz '2001-01-03 01:00:00+00',
                  interval '1' hour) as t(origin)
order by origin
;
```
```
┌────────────────────────┬─────────────────────┐
│          tstz          │         ts          │
├────────────────────────┼─────────────────────┤
│ 2023-10-07 00:00:00+00 │ 2023-10-07 00:00:00 │
│ 2023-10-06 01:00:00+00 │ 2023-10-07 01:00:00 │
└────────────────────────┴─────────────────────┘
```

### OS:

Ubuntu 22.04.3 LTS x64

### DuckDB Version:

v0.9.0 

### DuckDB Client:

cli

### Full Name:

rob m

### Affiliation:

unaffiliated

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by ghost**
Fixes #9260

the calculations used for binning a timestamptz to days was incorrect sometimes -- notably when the origin was not day aligned.

this change updates the calculation to do the standard 1d binning operation around a central origin.

when the timestamptz is greater than the origin the correct bin is calculated as follows.

```
//when days is positive this works.
days = icu_calc_days(origin, timestamptz); // returns days between origin and timestamptz.
bin = (days / bucket_width) * bucket_width;
``` 

when the timestamptz is less than the origin, and bin <= 0, we need to implement a floor-like operation to get the next bin less than or equal to the timestamptz. the calculation above provides the next bin greater to or equal so we need to adjust the bin.

```
if (timestamptz < icu_add_days_to_timetamptz(origin, bin)) {
  assert(bin <= 0 && timestamptz < origin);
  bin -= bucket_width;
}
assert(icu_add_days_to_timetamptz(origin, bin) <= timestamptz);
```

testing: added tests that trigger the bug and validated output against timescaledb. all unittests pass using a relassert build after change.

### CODE REVIEW

**Comment by robert-manchester** [comment]
@hawkfish

File: test/sql/function/timestamp/test_icu_time_bucket_timestamptz.test [hunk 1]
```diff
@@ -1213,3 +1213,40 @@ with t2 as (select '2000-10-28 00:00:00'::timestamptz + (n::text || 'hours')::in
 2000-10-30 17:00:00-08	2000-10-30 17:00:00-08
 2000-10-30 20:00:00-08	2000-10-30 20:00:00-08
 2000-10-30 23:00:00-08	2000-10-30 23:00:00-08
+
```
Comment (hawkfish) [inline test/sql/function/timestamp/test_icu_time_bucket_timestamptz.test:1216]: This is great but can we add all the tests from our discussion? I'm really paranoid now.  And can you add some from another TZ?

Comment (robert-manchester) [reply test/sql/function/timestamp/test_icu_time_bucket_timestamptz.test:1216]: added a couple more -- FWIW i am copy / pasting the output of the queries from timescaledb into the expected output for the test. so far so good.

Comment (robert-manchester) [reply test/sql/function/timestamp/test_icu_time_bucket_timestamptz.test:1216]: added this one:
```sql
select
  bool_and(
    date_part('epoch', time_bucket(interval '3' day, timestamptz '2023-06-07 16:08:09+00', origin))
    = date_part('epoch', time_bucket(interval '3' day, timestamp   '2023-06-07 16:08:09', origin at time zone 'UTC'))
  )
from
  generate_series(timestamptz '2023-01-03 00:00:00+05',
                  timestamptz '2024-01-04 00:00:00+05',
                  interval '7877' minute) as t(origin)
;
```
which makes sure that the icu and non-icu versions are equal.

**Review by hawkfish — APPROVED** [review]
For future reference, the brutal non-hours time zones are Newfoundland, Kathmandu and the Chatham Islands(!)

**Comment by robert-manchester** [comment]
> For future reference, the brutal non-hours time zones are Newfoundland, Kathmandu and the Chatham Islands(!)

hmm -- i thought India was good (5:30) but TIL there is no daylight savings time or other seasonal adjustments.

**Comment by robert-manchester** [comment]
R-CMD-check failure seems unrelated.

**Comment by hawkfish** [comment]
> > For future reference, the brutal non-hours time zones are Newfoundland, Kathmandu and the Chatham Islands(!)
> 
> hmm -- i thought India was good (5:30) but TIL there is no daylight savings time or other seasonal adjustments.

Oh it is - that's why I had an exclamation point! Mostly I'm just kicking myself because I was in NZ last month and the Chatham Islands are +12:15, which would be an interesting stress test of the TZ calculations.

**Comment by hawkfish** [comment]
@Mytherin Yes, I approved, but then started bantering about nasty TZs for testing. Sorry for the confusion!

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**